go-runtime/encoding/encoding.go

```diff
@@ -198,3 +201,219 @@ func encodeString(v reflect.Value, w *bytes.Buffer) error {
+		}
+		return dec.UnmarshalText([]byte(s))
+	}
+
+	switch v.Kind() {
+	case reflect.Struct:
+		return decodeStruct(data, v)
+
+	case reflect.Ptr:
+		if v.IsNil() {
+			v.Set(reflect.New(v.Type().Elem()))
+		}
+		return decodeValue(data, v.Elem())
+
+	case reflect.Slice:
+		if v.Type().Elem().Kind() == reflect.Uint8 {
+			return decodeBytes(data, v)
+		}
+		return decodeSlice(data, v)
+
+	case reflect.Map:
+		return decodeMap(data, v)
+
+	case reflect.String:
+		return decodeString(data, v)
+
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return decodeInt(data, v)
+
+	case reflect.Float32, reflect.Float64:
+		return decodeFloat(data, v)
+
+	case reflect.Bool:
+		return decodeBool(data, v)
+
+	case reflect.Interface:
+		if v.Type().NumMethod() != 0 {
+			return fmt.Errorf("the only interface type supported is any, not %s", v.Type())
+		}
+
+		var anyInterface any
+		if err := json.Unmarshal(data, &anyInterface); err != nil {
+			return err
+		}
+		reflectedValue := reflect.ValueOf(anyInterface)
+		if !reflectedValue.Type().AssignableTo(v.Type()) {
+			return fmt.Errorf("cannot assign type %s to any interface", reflectedValue.Type())
+		}
+		v.Set(reflectedValue)
+		return nil
+
+	default:
+		return fmt.Errorf("unsupported type: %s", v.Type())
+	}
+}
+
+func decodeStruct(data []byte, v reflect.Value) error {
+	var jsonData map[string]json.RawMessage
+	if err := json.Unmarshal(data, &jsonData); err != nil {
+		return err
+	}
+
+	for i := 0; i < v.NumField(); i++ {
+		field := v.Field(i)
+		fieldType := v.Type().Field(i)
+		jsonKey := strcase.ToLowerCamel(fieldType.Name)
+		if _, ok := jsonData[jsonKey]; ok {
+			fieldTypeStr := fieldType.Type.String()
+			switch {
+			case fieldTypeStr == "*Unit" || fieldTypeStr == "Unit":
+				if fieldTypeStr == "*Unit" && field.IsNil() {
+					field.Set(reflect.New(field.Type().Elem()))
+				}
+
+			default:
+				if err := decodeValue(jsonData[jsonKey], field); err != nil {
+					return fmt.Errorf("error decoding field '%s': %w", fieldType.Name, err)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func decodeString(data []byte, v reflect.Value) error {
+	var s string
+	if err := json.Unmarshal(data, &s); err != nil {
+		return err
+	}
+	v.SetString(s)
+	return nil
+}
+
+func decodeInt(data []byte, v reflect.Value) error {
+	var n int64
+	n, err := strconv.ParseInt(string(data), 10, 64)
+	if err != nil {
+		return err
+	}
+	v.SetInt(n)
+	return nil
+}
+
+func decodeFloat(data []byte, v reflect.Value) error {
+	var f float64
+	if err := json.Unmarshal(data, &f); err != nil {
+		return err
+	}
+	v.SetFloat(f)
+	return nil
+}
+
+func decodeBool(data []byte, v reflect.Value) error {
+	var b bool
+	if err := json.Unmarshal(data, &b); err != nil {
+		return err
+	}
+	v.SetBool(b)
+	return nil
+}
+
+func decodeBytes(data []byte, v reflect.Value) error {
+	var s string
+	if err := json.Unmarshal(data, &s); err != nil {
+		return err
+	}
+	b, err := base64.StdEncoding.DecodeString(s)
```

Comment:
Can't decode directly into []byte and get this for free?

Reply:
oop yes